### PR TITLE
Variable Clocks Model 12.0 Test Run

### PR DESCRIPTION
**Pull request by patricktnast**
### Description

- *Category*: 
implementation
- *JIRA issue*: [MIC-4644](https://jira.ihme.washington.edu/browse/MIC-4644)
- *Research reference*: 
https://vivarium-research.readthedocs.io/en/latest/models/concept_models/vivarium_nutrition_optimization/kids/concept_model.html#variable-timestep-rules
### Changes and notes
 
* Added population and disease model modifiers of simiu;ant clocks
* Moved dead simulants to end of simulation
### Verification and Testing

Ran simulation

### CODE REVIEW

File: src/vivarium_gates_nutrition_optimization_child/components/fertility.py [hunk 1]
```diff
@@ -53,7 +54,8 @@ def _get_birth_records(builder: Builder) -> pd.DataFrame:
 
         file_path = data_directory / f"scenario_{scenario}_draw_{draw}_seed_{seed}.hdf"
         birth_records = pd.read_hdf(file_path)
-        birth_records["birth_date"] = pd.to_datetime(birth_records["birth_date"])
+        # birth_records["birth_date"] = pd.to_datetime(birth_records["birth_date"])
+        birth_records["birth_date"] = pd.to_datetime("2024-12-31 12:00:01")
```
Comment (patricktnast) [inline src/vivarium_gates_nutrition_optimization_child/components/fertility.py:58]: This was to keep the step_backward() consistent; it causes a bug otherwise.

File: src/vivarium_gates_nutrition_optimization_child/components/population.py [hunk 2]
```diff
@@ -104,6 +109,22 @@ def on_initialize_simulants(self, pop_data: SimulantData) -> None:
     def _get_location(self, builder: Builder) -> str:
         return builder.data.load("population.location")
 
+    def modify_step(self, index: pd.Index) -> pd.Series:
+        """
+        Sets simlant step size to 0.5 for neonates and 4 for 1-5 months.
+        """
+        neonates = self.population_view.get(index, f"age < {NEONATAL_END_AGE}").index
+        one_to_five_mos = self.population_view.get(
+            index, f"age >= {NEONATAL_END_AGE} and age < 0.416667"
```
Comment (stevebachmeier) [inline src/vivarium_gates_nutrition_optimization_child/components/population.py:118]: probably want to make a constant out of 0.416667

File: src/vivarium_gates_nutrition_optimization_child/components/population.py [hunk 3]
```diff
@@ -104,6 +109,22 @@ def on_initialize_simulants(self, pop_data: SimulantData) -> None:
...
+        ).index
+        step_size = pd.concat(
+            [
+                pd.Series(pd.Timedelta(days=0.5), index=neonates),
```
Comment (stevebachmeier) [inline src/vivarium_gates_nutrition_optimization_child/components/population.py:122]: Shouldn't these Timedelta values be coming out of the model spec?

File: src/vivarium_gates_nutrition_optimization_child/components/fertility.py (hunk 1 above)
Comment (stevebachmeier) [reply src/vivarium_gates_nutrition_optimization_child/components/fertility.py:58]: If this is a constant it should probably be stored somewhere.

File: src/vivarium_gates_nutrition_optimization_child/components/causes.py [hunk 4]
```diff
@@ -1,17 +1,81 @@
...
     RiskAttributableDisease as RiskAttributableDisease_,
 )
 from vivarium_public_health.disease import SusceptibleState
 from vivarium_public_health.disease.transition import TransitionString
 
 
-def SIS_with_birth_prevalence(cause: str) -> DiseaseModel:
+class DiseaseModel(DiseaseModel_):
+    @property
+    def columns_required(self):
```
Comment (stevebachmeier) [inline src/vivarium_gates_nutrition_optimization_child/components/causes.py:17]: Are these not already defined in the super DiseaseModel?

File: src/vivarium_gates_nutrition_optimization_child/components/causes.py [hunk 5]
```diff
@@ -1,17 +1,81 @@
...
+    healthy = SusceptibleState(cause, allow_self_transition=True)
+    infected = DiseaseState(cause, allow_self_transition=True)
+
+    healthy.add_rate_transition(infected)
+    infected.add_rate_transition(healthy)
+
+    infected_step_size_function = (
+        {"variable_step_sizes": {infected.state_id: float(step_size_days)}}
+        if step_size_days
+        else {}
```
Comment (stevebachmeier) [inline src/vivarium_gates_nutrition_optimization_child/components/causes.py:46]: nit: maybe should be None

File: src/vivarium_gates_nutrition_optimization_child/components/causes.py [hunk 6]
```diff
@@ -1,17 +1,81 @@
...
+        allow_self_transition=True,
+    )
+
+    healthy.add_rate_transition(infected)
+    infected.add_dwell_time_transition(healthy)
+
+    infected_step_size_function = (
+        {"variable_step_sizes": {infected.state_id: float(step_size_days)}}
+        if step_size_days
+        else {}
```
Comment (stevebachmeier) [inline src/vivarium_gates_nutrition_optimization_child/components/causes.py:70]: and here

File: src/vivarium_gates_nutrition_optimization_child/components/causes.py [hunk 7]
```diff
@@ -1,17 +1,81 @@
...
+        cause,
+        get_data_functions={"dwell_time": lambda _, __: duration},
+        allow_self_transition=True,
+    )
+
+    healthy.add_rate_transition(infected)
+    infected.add_dwell_time_transition(healthy)
+
+    infected_step_size_function = (
+        {"variable_step_sizes": {infected.state_id: float(step_size_days)}}
```
Comment (stevebachmeier) [inline src/vivarium_gates_nutrition_optimization_child/components/causes.py:68]: this doesn't match the typing in DiseaseModel (https://github.com/ihmeuw/vivarium_public_health/blob/main/src/vivarium_public_health/disease/model.py#L68).

Shouldn't the values be a callable?

File: src/vivarium_gates_nutrition_optimization_child/components/causes.py [hunk 8]
```diff
@@ -1,17 +1,81 @@
 """
 Component to include birth prevalence in SIS model.
```
Comment (rmudambi) [inline src/vivarium_gates_nutrition_optimization_child/components/causes.py:2]: Nit: this description is out of date now

File: src/vivarium_gates_nutrition_optimization_child/components/causes.py [hunk 9]
```diff
@@ -1,17 +1,81 @@
...
 )
 from vivarium_public_health.disease import SusceptibleState
 from vivarium_public_health.disease.transition import TransitionString
 
 
-def SIS_with_birth_prevalence(cause: str) -> DiseaseModel:
+class DiseaseModel(DiseaseModel_):
+    @property
+    def columns_required(self):
+        return ["age", "sex", "alive", "tracked"]
```
Comment (rmudambi) [inline src/vivarium_gates_nutrition_optimization_child/components/causes.py:18]: Maybe use a call to `super().columns_required` to make it clearer what was added.

File: src/vivarium_gates_nutrition_optimization_child/components/causes.py [hunk 10]
```diff
@@ -1,17 +1,81 @@
...
+
+    def modify_step(self, state, step_size, index):
+        infected = self.population_view.get(
+            index,
+            f"{self.state_column} == '{state}' and alive == 'alive' and tracked == True",
+        ).index
+        return pd.Series(pd.Timedelta(days=step_size), index=infected)
+
+
+def SIS(cause: str, step_size_days: str = None) -> DiseaseModel:
```
Comment (rmudambi) [inline src/vivarium_gates_nutrition_optimization_child/components/causes.py:36]: In the interest of wrapping up this model, this is fine for now, but a better approach would be to be specifying step-size values as configuration values and having that code inside `DiseaseModel` itself rather than as an argument to the various implementations. That would reduce code duplication, make the configuration easier to understand, and most significantly enable a user to set a step-size for any state in any disease model. 

That being said, if we are moving towards a method of determining step-sizes that is more sophisticated than having the user define them all individually, this approach will need to be revisited.

File: src/vivarium_gates_nutrition_optimization_child/components/population.py [hunk 11]
```diff
@@ -19,8 +19,12 @@
     assign_demographic_proportions,
     load_population_structure,
 )
+from vivarium_public_health.population.mortality import Mortality as _Mortality
```
Comment (rmudambi) [inline src/vivarium_gates_nutrition_optimization_child/components/population.py:22]: Nit:
```suggestion
from vivarium_public_health.population.mortality import Mortality as Mortality_
```

File: src/vivarium_gates_nutrition_optimization_child/components/population.py [hunk 12]
```diff
@@ -133,3 +154,15 @@ def on_initialize_simulants(self, pop_data: SimulantData) -> None:
         population.loc[population.index % 2 == 1, "sex"] = "Male"
         self.register_simulants(population[list(self.key_columns)])
         self.population_view.update(population)
+
+
+class Mortality(_Mortality):
```
Comment (rmudambi) [inline src/vivarium_gates_nutrition_optimization_child/components/population.py:159]: Nit: 
```suggestion
class Mortality(Mortality_):
```

File: src/vivarium_gates_nutrition_optimization_child/components/population.py [hunk 13]
```diff
@@ -104,6 +109,22 @@ def on_initialize_simulants(self, pop_data: SimulantData) -> None:
     def _get_location(self, builder: Builder) -> str:
         return builder.data.load("population.location")
 
+    def modify_step(self, index: pd.Index) -> pd.Series:
+        """
+        Sets simlant step size to 0.5 for neonates and 4 for 1-5 months.
```
Comment (rmudambi) [inline src/vivarium_gates_nutrition_optimization_child/components/population.py:114]: Typo: 
```suggestion
        Sets simulant step size to 0.5 for neonates and 4 for 1-5 months.
```

File: src/vivarium_gates_nutrition_optimization_child/components/population.py [hunk 14]
```diff
@@ -77,6 +81,7 @@ def setup(self, builder: Builder) -> None:
 
         self.start_time = get_time_stamp(builder.configuration.time.start)
         self.location = self._get_location(builder)
+        builder.time.register_step_modifier(self.modify_step)
```
Comment (rmudambi) [inline src/vivarium_gates_nutrition_optimization_child/components/population.py:84]: Out of scope for now, but I wonder if we should make `modify_step` a "special" function name for `Component` objects like `on_time_step`, etc. By this I mean have component's `setup_component` function look for this method and automatically call `builder.time.register_step_modifier` so the component implementer doesn't need to remember to do it (and remove boilerplate).

File: src/vivarium_gates_nutrition_optimization_child/components/population.py [hunk 15]
```diff
@@ -133,3 +154,15 @@ def on_initialize_simulants(self, pop_data: SimulantData) -> None:
         population.loc[population.index % 2 == 1, "sex"] = "Male"
         self.register_simulants(population[list(self.key_columns)])
         self.population_view.update(population)
+
+
+class Mortality(_Mortality):
+    def setup(self, builder):
```
Comment (rmudambi) [inline src/vivarium_gates_nutrition_optimization_child/components/population.py:160]: Nit: typing here and below

File: src/vivarium_gates_nutrition_optimization_child/constants/data_keys.py [hunk 16]
```diff
@@ -466,6 +466,7 @@ class CGFCategories(Enum):
     MILD = "mild"
     BETTER_MODERATE = "better_moderate"
     WORSE_MODERATE = "worse_moderate"
+    # MODERATE = "moderate"
```
Comment (rmudambi) [inline src/vivarium_gates_nutrition_optimization_child/constants/data_keys.py:469]: Is this supposed to be here?

File: src/vivarium_gates_nutrition_optimization_child/model_specifications/nutrition_optimization_child.yaml [hunk 17]
```diff
@@ -54,25 +48,29 @@ components:
...
             - RiskAttributableDisease('cause.moderate_protein_energy_malnutrition', 'risk_factor.child_wasting')
             - RiskAttributableDisease('cause.severe_protein_energy_malnutrition', 'risk_factor.child_wasting')
+            - Mortality()
             - MortalityObserver()
             - ResultsStratifier()
 
 configuration:
     input_data:
-        input_draw_number: 0
+        input_draw_number: 602
```
Comment (rmudambi) [inline src/vivarium_gates_nutrition_optimization_child/model_specifications/nutrition_optimization_child.yaml:63]: Are these changes intended to be here?

File: src/vivarium_gates_nutrition_optimization_child/components/causes.py (hunk 7 above)
Comment (rmudambi) [reply src/vivarium_gates_nutrition_optimization_child/components/causes.py:68]: Good catch. This should probably be done, though it's not critical, since this implementation is likely temporary.

**Review by rmudambi — APPROVED** [review]

File: src/vivarium_gates_nutrition_optimization_child/components/causes.py (hunk 4 above)
Comment (patricktnast) [reply src/vivarium_gates_nutrition_optimization_child/components/causes.py:17]: cf. Rajan's comment, I added two columns. It's clearer to make the call explicitly to super.columns_required

File: src/vivarium_gates_nutrition_optimization_child/model_specifications/nutrition_optimization_child.yaml [hunk 18]
```diff
@@ -54,25 +48,29 @@ components:
...
             - MMSEffectOnGestationalAge()
             - BirthWeightShiftEffect()
-            - SIS_with_birth_prevalence('diarrheal_diseases')
+            - SIS_with_birth_prevalence('diarrheal_diseases', '4.0')
+            - SIS_fixed_duration('measles', '10.0', '4.0')
+            - SIS('lower_respiratory_infections', '4.0')
+            - SIS('malaria', '4.0')
             - RiskAttributableDisease('cause.moderate_protein_energy_malnutrition', 'risk_factor.child_wasting')
             - RiskAttributableDisease('cause.severe_protein_energy_malnutrition', 'risk_factor.child_wasting')
+            - Mortality()
```
[the remark this answers is not in the document]
Comment (patricktnast) [reply src/vivarium_gates_nutrition_optimization_child/model_specifications/nutrition_optimization_child.yaml:57]: Probably not quite in its current form; but the idea in any case is that implementing it in model first is a sort of QA / proving ground for integrating the more abstract idea in VPH, yeah

File: src/vivarium_gates_nutrition_optimization_child/components/population.py (hunk 3 above)
Comment (patricktnast) [reply src/vivarium_gates_nutrition_optimization_child/components/population.py:122]: If you mean that they are related to the min / standard step sizes somehow, I'd say not. These are hard-coded values specific to this "test run" implementation; but in principle you could make them configurable from the model spec ofc. In the long run, these values will be replaced with more general expressions / equations

File: src/vivarium_gates_nutrition_optimization_child/components/fertility.py (hunk 1 above)
Comment (patricktnast) [reply src/vivarium_gates_nutrition_optimization_child/components/fertility.py:58]: it's not necessarily a constant, but it makes more sense to pull the information directly from the configuration I think

File: src/vivarium_gates_nutrition_optimization_child/components/causes.py (hunk 7 above)
Comment (patricktnast) [reply src/vivarium_gates_nutrition_optimization_child/components/causes.py:68]: i fixed this in a lazy way (lambda'ing the dictionary) in the interest of coming back to the details here after the new implementation of variable clocks is in place

**Review by albrja — APPROVED** [review]

**Review by rmudambi — APPROVED** [review]
